src/bom/extract.rs

```diff
@@ -102,45 +84,32 @@ pub fn do_bitcode_extraction(extract_options : &ExtractOptions,
     //
     // NOTE: Ideally, we would be able to use the tar library for this instead
     // of calling out to tar.
-    let mut tar_args = Vec::new();
-    tar_args.push(OsString::from("xif"));
-    tar_args.push(OsString::from(ok_tar_name));
-    match Command::new("tar").args(&tar_args).current_dir(tmp_path).spawn() {
-        Err(msg) => {
-            return Err(anyhow::Error::new(ExtractError::ErrorRunningCommand(String::from("tar"), tar_args, msg)));
-        }
-        Ok(mut child) => {
-            let _rc = child.wait();
-        }
-    }
+
+    extract_ops.push_op(
+        &SubProcOperation::new(
+            &Executable::new("tar",
+                             ExeFileSpec::Append,
+                             ExeFileSpec::NoFileUsed))
+            .set_label("tar:unpack")
+            .push_arg("xif")
+            .set_dir(tmp_path)  // Extracted files will be placed here
+            // Input here is explicitly the section dump target file rather
+            // than the output of the previous op.
+            .set_input_file(&FileArg::loc(tar_path))
+    );
 
     // Now all the files extracted from bitcode.tar into tmp_dir should be linked
     // together to create the final bitcode file.
 
-    let mut llvm_link_args = Vec::new();
-    llvm_link_args.push(OsString::from("-o"));
-    llvm_link_args.push(OsString::from(&extract_options.output));
-
-    let mut bc_glob = String::new();
-    bc_glob.push_str(&OsString::from(tmp_path).into_string().unwrap());
-    bc_glob.push_str("/*.bc");
-    let bc_files = glob::glob(&bc_glob)?;
-    for bc_entry in bc_files {
-        let bc_file = bc_entry?;
-        debug!("#: bitcode file: {:?}", bc_file.file_name().unwrap());
-        llvm_link_args.push(OsString::from(bc_file));
-    }
-
-    let llvm_link = OsString::from(extract_options.llvm_link_path.as_ref().unwrap_or(&String::from("llvm-link")));
-    match Command::new(&llvm_link).args(&llvm_link_args).spawn() {
-        Err(msg) => {
-            let llvm_link_str = llvm_link.into_string().unwrap();
-            return Err(anyhow::Error::new(ExtractError::ErrorRunningCommand(llvm_link_str, llvm_link_args, msg)));
-        }
-        Ok(mut child) => {
-            let _rc = child.wait();
-        }
-    }
-
-    Ok(0)
+    let linker = String::from(extract_options.llvm_link_path
+                              .as_ref().unwrap_or(&"llvm-link"
+                                                  .to_string()));
+    let mut link_bc_files = extract_ops.push_op(
+        &SubProcOperation::new(&Executable::new(&linker,
+                                                ExeFileSpec::Append,
+                                                ExeFileSpec::option("-o"))));
+    link_bc_files.set_input_file(&FileArg::glob_in(tmp_path, "*.bc"));
+
+    let executor = get_executor(extract_options.verbose.len());
+    extract_ops.execute_here(&executor).map(|_| 0)
```

Comment:
Just following up on the previous question, how do errors get reported here? What happens if the objcopy fails? Does this `map` just make that into a success by returning 0? Or was it turned into some other `Err` inside of the process executor?

Reply:
Errors from `execute_here` still get returned from `do_bitcode_extraction` via the `anyhow::Result<i32>`.  The `map` just converts the success result from `execute_here` into the success result for `do_bitcode_extraction` but doesn't touch `Err` values (see https://doc.rust-lang.org/std/result/enum.Result.html#method.map).

The `push_op` calls add operations to a declared chain of operations, then the `execute_here` actually performs the operations in sequence, aborting with error if any operation fails (like the `objcopy`).  For more details, see the paragraph before and after the 4th code example at https://kquick.github.io/chainsop/chainsop/index.html and https://kquick.github.io/chainsop/chainsop/struct.ChainedOps.html#method.push_op.